File: inflate.cpp
/*
ID: totheso1
LANG: C++14
TASK: inflate
*/

#include <algorithm>
#include <fstream>
#include <iterator>
#include <vector>

using namespace std;

static ifstream fin("inflate.in");
static ofstream fout("inflate.out");

#define endl '\n'

template <typename T>
T fin_get() {
  return *istream_iterator<T>(fin);
}

template <typename C>
C fin_get_collection(const int size) {
  C ret;
  copy_n(istream_iterator<typename C::value_type>(fin), size,
         back_inserter(ret));
  return ret;
}

static const int TIME_TOTAL = fin_get<int>();
static const int CATEGORY_COUNT = fin_get<int>();
static vector<int> max_point(TIME_TOTAL + 1, 0);

int main() {
  for (int point, minute; fin >> point >> minute;) {
    for (auto b = begin(max_point) + minute, e = end(max_point); b < e; ++b) {
      *b = max(*b, *(b - minute) + point);
    }
  }
  fout << max_point.back() << endl;
}

File: inflate.py
"""
ID: totheso1
LANG: PYTHON3
TASK: inflate
NOTE:
    Python3 will get Time Limit Exceeded.
    Please use C++, Java, Pascal instead.
"""

from typing import Any

fin = open("inflate.in", "r")
fout = open("inflate.out", "w")


def fprint(*args: Any, **kwargs: Any) -> None:
    print(*args, file=fout, **kwargs)


# Code start
TIME_TOTAL, CATEGORY_COUNT = map(int, fin.readline().split())
max_point = [0] * (TIME_TOTAL + 1)
for line in fin:
    point, minute = map(int, line.split())
    for i in range(minute, TIME_TOTAL + 1):
        max_point[i] = max(max_point[i], max_point[i - minute] + point)
fprint(max_point[-1])
# Code end
